### PR TITLE
Add poly1305 and x25519 crypto primitives

### PR DESCRIPTION
**Pull request by tiehuis**
@shawnl, @kristate 

Sorry if you happened to be working on this. I got side-tracked and ended up just implementing it all so I thought I'd push early.

The main things remaining are:
 - [x] Run through and do a clean up
 - [x] Make the API more idiomatic for Zig
 - [x] Quick scan for performance improvements where possible

These are translated from [monocypher](https://monocypher.org/) which
has fairly competitive performance while remaining quite simple.

Initial performance comparision:

```
Zig:
 Poly1305: 1423 MiB/s
 X25519:   8671 exchanges per second
```

```
Monocypher:
 Poly1305: 1567 MiB/s
 X25519:   10539 exchanges per second
```

There is room for improvement and no real effort has been made at all in
optimization beyond a direct translation.

### CODE REVIEW

**Comment by tiehuis** [comment]
I used this programs for benchmarking: https://gist.github.com/tiehuis/185c975c4c07e465ab72c68abf11948a

**Comment by shawnl** [comment]
> Make the API more idiomatic for Zig

Yeah, this was my big gripe in looking at your code. You also have to write the IETF aead construction for chacha20-poly1305

**Comment by tiehuis** [comment]
Alright the API across all functions is a bit more consistent now. I've generalized the Mac functionality of Poly1305 to match the Hmac implementations and X25519 now is encapsulated in its own struct.

The interface for crypto-related functions is this at the moment:

```
const Mac = interface {
    pub const mac_length: comptime_int;

    // different requirements, e.g. HMAC construction allows variable sizes whilst Poly1305 is fixed 32-byte.
    pub const minimum_key_length: comptime_int;

    // Caller must ensure:
    //   - out.len >= Mac.length
    //   - key.len >= Mac.minimum_key_length
    pub fn create(out: []u8, msg: []const u8, key: []const u8) void;

    // Caller must ensure:
    //   - key.len is sufficient size.
    pub fn init(key: []const u8) Mac;

    pub fn update(self: *Self, msg: []const u8) void;

    pub fn final(self: *Self, out: []u8);
};

test "mac" {
    const key = "abcdefg";
    const m = Mac.init(key);

    m.update("message");
    m.update("message pt 2");

    var mac: [Mac.mac_length]u8 = undefined;
    m.final(mac[0..]);
}


const Hash = interface {
    pub const digest_length: comptime_int;

    pub const block_length: comptime_int;

    // Caller must ensure:
    //   - out.len >= length
    pub fn hash(out: []u8, msg: []const u8) void;

    pub fn init() Hash;

    pub fn update(self: *Self, msg: []const u8) void;

    pub fn final(self: *Self, out: []u8);
};

test "hash" {
    const key = "abcdefg";
    const h = Hash.init(key);

    h.update("message");
    h.update("message pt 2");

    var hash: [Hash.digest_length]u8 = undefined;
    h.final(hash[0..]);
}


const DhKeyExchange = interface {
    pub const secret_length: comptime_int;

    // Minimum key length for private/public keys. Assume private.len == public.len
    // always (may be schemes which don't obey this in future?)
    pub const minimum_key_length: comptime_int;

    // Caller must ensure
    //   - private_key.len >= DhKeyExchange.minimum_key_length
    //   - public_key.len  >= DhKeyExchange.minimum_key_length
    //
    // Returns true if the public key was malicious.
    pub fn create(out: []u8, private_key: []const u8, public_key: []const u8) bool ;

    // Generate the corresponding public key given a private key.
    pub fn createPublicKey(public_key: []const u8, private_key: []const u8) bool;
};

test "dh" {
    const private = "abdefg";
    const public = "hijklm";

    var shared_secret: [DhKeyExchange.secret_length]u8 = undefined;
    DhKeyExchange.create(shared_secret[0..], private, public);
}
```

I'm not going to implement an AEAD interface atop this in this PR. That is pretty trivial once this is merged however.

File: std/crypto/hmac.zig [hunk 1]
```diff
@@ -7,75 +7,92 @@ pub const HmacMd5 = Hmac(crypto.Md5);
...
-            debug.assert(output.len >= H.digest_size);
-            debug.assert(H.digest_size <= H.block_size); // HMAC makes this assumption
-            var scratch: [H.block_size]u8 = undefined;
+        o_key_pad: [Hash.block_length]u8,
+        i_key_pad: [Hash.block_length]u8,
+        scratch: [Hash.block_length]u8,
+        hash: Hash,
+
+        // HMAC(k, m) = H(o_key_pad | H(i_key_pad | msg)) where | is concatenation
+        pub fn create(out: []u8, msg: []const u8, key: []const u8) void {
```
Comment (tiehuis) [inline std/crypto/hmac.zig]: Feel free to suggest a better name than `create` here. I've used this for key exchange as well however and given the type is known as a Mac it makes enough sense to me.

File: std/crypto/throughput_test.zig [hunk 2]
```diff
@@ -1,38 +1,198 @@
...
+            usage();
+            return;
+        } else {
+            usage();
+            std.os.exit(1);
+        }
+    }
+
+    inline for (hashes) |H| {
+        // TODO: Inverted early continue case here segfaults compiler. Create reduced test case.
```
Comment (tiehuis) [inline std/crypto/throughput_test.zig]: This needs a reduced test case but I couldn't get an immediately smaller one. I'll make an issue linking to this for the moment if I don't have a reduction shortly (and the issue isn't already filed).

```#0  0x0000555558b786d0 in llvm::BasicBlock::getContext() const ()
#1  0x0000555558c083e9 in llvm::BranchInst::BranchInst(llvm::BasicBlock*, llvm::BasicBlock*, llvm::Value*, llvm::Instruction*) ()
#2  0x0000555558ba3208 in LLVMBuildCondBr ()
#3  0x0000555555efd84a in ir_render_instruction(CodeGen*, IrExecutable*, IrInstruction*) [clone .isra.182] ()
#4  0x0000555555f03a0e in codegen_build(CodeGen*) ()
#5  0x0000555555e49731 in main ()
```

**Comment by tiehuis** [comment]
Think this is in a good enough state right now to have a look.

**Pull request merged**